### DISCUSSION TITLE
Asynchronous listeners can cause some problems

### DISCUSSION THREAD

**Discussion by pichoemr** [lenra-io/lenra_cli#113] [General]
As mentioned in this issue https://github.com/lenra-io/application-runner/issues/178, we currently have some issues with the listeners especially OnUserFirstJoin & OnSessionStart. Here is the current environment and session architecture:

```mermaid
graph
Environment.supervisor --> Environment.Managers
Environment.Managers --> Environment.Manager
Session.Supervisor --> Session.Managers
Session.Managers --> Session.Manager
Session.Manager --> EventHandler
Environment.Manager --> EventHandler
```

When the session's Genserver starts, we wait for the environment to start and the OnEnvStart listeners to end. After the environment starts, the session sends the OnUserFirstJoin & OnSessionStart listeners asynchronously, which is a problem because we are building the interface but the listeners are not finished, so if we need the data created by these listeners, we get an error.

In fact, we can't synchronize the OnUserFirstJoin and OnSessionStart listeners because the application api requests have a token, to validate the token, we need to get the token saved by the session token genserver. But to call these token genserver, we need to get the pid by calling fetch_module_pid! function, which itself calls the session to get the supervisor's PID. So in fact, we can't synchronize the Listeners

@lenra-io/devs

**Comment by taorepoara** [comment]
I think that system events like those should not be async. This make the app waiting for them, but I don't think that it's a problem since they already wait for the OnEnvStart one.

**Comment by Nesqwik** [comment]
This issue have probably been fixed by the new application_runner using mongo.

**Comment by taorepoara** [comment]
@pichoemr is that solved ?

**Reply by pichoemr** [reply]
@taorepoara Normally yes, OnUserFirstJoin and OnSessionStart are now synchronized.
